src/Shared/Debugging/PrintLineDebugger.cs

```diff
@@ -0,0 +1,209 @@
+            });
+
+        public static Lazy<PrintLineDebugger> Default =
+            new Lazy<PrintLineDebugger>(() => Create(null, null, false));
+
+        public static Lazy<PrintLineDebugger> DefaultWithProcessInfo =
+            new Lazy<PrintLineDebugger>(() => Create(null, null, true));
+
+        private static readonly Lazy<NodeMode> ProcessNodeMode = new Lazy<NodeMode>(
+            () =>
+            {
+                return ScanNodeMode(Environment.CommandLine);
+
+                NodeMode ScanNodeMode(string input)
+                {
+                    var match = Regex.Match(input, @"/nodemode:(?<nodemode>[12\s])(\s|$)", RegexOptions.IgnoreCase);
+
+                    if (!match.Success)
+                    {
+                        return NodeMode.CentralNode;
+                    }
+                    var nodeMode = match.Groups["nodemode"].Value;
+
+                    Trace.Assert(!string.IsNullOrEmpty(nodeMode));
+
+                    switch (nodeMode)
+                    {
+                        case "1":
+                            return NodeMode.OutOfProcNode;
+                        case "2":
+                            return NodeMode.OutOfProcTaskHostNode;
+                        default:
+                            throw new NotImplementedException();
+                    }
+                }
+            });
+
+        private readonly string _id;
+
+        private readonly CommonWriterType _writerSetByThisInstance;
+
+        public static string ProcessInfo
+            =>
+                $"{ProcessNodeMode.Value}_PID={Process.GetCurrentProcess() .Id}({Process.GetCurrentProcess() .ProcessName})x{(Environment.Is64BitProcess ? "64" : "86")}"
+            ;
+
+        public PrintLineDebugger(string id, CommonWriterType writer)
+        {
+            _id = id ?? string.Empty;
+
+            if (writer != null)
+            {
+                SetWriter(writer);
+
+                // we wrap the original writer with a locking writer in SetWriter, so get the actual writer that was set
+                _writerSetByThisInstance = GetStaticWriter();
+            }
+        }
+
+        public void Dispose()
+        {
+            ReleaseUnmanagedResources();
+            GC.SuppressFinalize(this);
+        }
+
+        public static CommonWriterType GetStaticWriter()
+        {
+            return (CommonWriterType) CommonWriterProperty.Value.GetValue(null, null);
+            ;
+        }
+
+        public static void SetWriter(CommonWriterType writer)
+        {
+            var currentWriter = GetStaticWriter();
+
+            if (currentWriter != null)
+            {
+                ErrorUtilities.ThrowInternalError("Cannot set a new writer over an old writer. Remove the old one first");
+            }
+
+            // wrap with a lock so multi threaded logging does not break messages apart
+            CommonWriterProperty.Value.SetValue(null, (CommonWriterType) LockWrappedWriter);
```

Comment:
Is it safe to set the value in this way? Is there a race between multiple threads hitting this line?

Reply:
The usage intent I had in mind was to investigate one problem at a time, which means setting one writer only once (either from a test, if tests are failing, or from the main method or BuildManager when investigating under VS). Hence, no need for thready safety. If you're alright with this assumption, I'll put in a comment about it.

Reply:
Sounds good.